modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
-import { Arr, Cell, Obj } from '@ephox/katamari';
+import { Arr, Cell, Obj, Optional, Type } from '@ephox/katamari';
 
 import Editor from '../api/Editor';
-import Tools from '../api/util/Tools';
+import { FormatVars } from './FormatTypes';
 import * as FormatUtils from './FormatUtils';
 import * as MatchFormat from './MatchFormat';
 
 export type FormatChangeCallback = (state: boolean, data: { node: Node; format: string; parents: any }) => void;
 
-export type FormatCallbacks = Record<string, FormatChangeCallback[]>;
+export type RegisteredFormats = Record<string, CallbackGroup>;
 
-export interface FormatData {
-  similar?: boolean;
-  callbacks: FormatChangeCallback[];
+export interface UnbindFormatChanged {
+  unbind: () => void;
 }
 
-export type RegisteredFormats = Record<string, FormatData>;
+interface CustomCallback {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly vars: FormatVars;
+  readonly callback: FormatChangeCallback;
+}
 
-export interface UnbindFormatChanged {
-  unbind: () => void;
+interface CallbackGroup {
```

Comment:
I'm not really a fan of this data structure, as it requires a lot of looping to do the various operations (remember we this is for NodeChange so performance does matter somewhat). I'll check this out and have a play locally to see if I can come up with any other suggestions.

Reply:
What it comes down to to me isn't so much the looping through the data structure that we have to do, but the number of times we need to walk through the parents array calling `formatter.matchNode`. Right now, that's set at one walk operation for each unique `(format, similar, vars)` combination that's been registered, which means we aren't duplicating any work at least.

I we could get different information out of `formatter.matchNode` (like getting it to tell us which variables are applied, instead of giving us a yes/no for a particular set of variables) it might be possible to cut that down to now walk operation for each unique `(format, similar)` combination, or maybe even just one walk operation per format — but with how the formatter is set up I don't think that sort of path is possible.

Anyway, if you've got time, and you can think of a way to cut down the number of walk operations that I've missed, let me know and I'll do what I can to fix it. This is 100% a hot code path, it'd be nice to make it snappy

Reply:
Yeah, it may not be possible to do anything here. I'm fine with merging this as is if we can't find anything.